File: leetcode-js/209-minimum-size-subarray-sum.js
/**
 * @param {number} s
 * @param {number[]} nums
 * @return {number}
 */
var minSubArrayLen = function(s, nums) {
    let left = 0;
    let right = 0;
    let sum = 0;
    let len = nums.length + 1;

    while (right < nums.length) {
        while (right < nums.length && sum < s) {
            sum += nums[right];
            right += 1;
        }
        while (left < right && sum >= s) {
            len = Math.min(len, right - left);
            sum -= nums[left];
            left += 1;
        }
    }

    return len === nums.length + 1 ? 0 : len;
};

console.log(minSubArrayLen(7, [2,3,1,2,8,-1]))

File: leetcode-java/209-minimum-size-subarray-sum.java
public class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        if (nums.length == 0) {
            return 0;
        }

        int i = 0;
        int j = 0;
        int sum = 0;
        int min = Integer.MAX_VALUE;

        while (j < nums.length) {
            sum += nums[j];
            j++;
            while (sum >= s) {
                min = Math.min(min, j - i);
                sum -= nums[i];
                i++;
            }
        }

        return min == Integer.MAX_VALUE ? 0 : min;
    }
}
